core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,319 @@ class DAGSchedulerSuite
+      complete(stage1Attempt,
+        Seq((FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null))
+          ++ stage1Successes
+      )
+
+      // this will (potentially) trigger a resubmission of stage 0, since we've lost some of its
+      // map output, for the next iteration through the loop
+      scheduler.resubmitFailedStages()
+
+      if (attempt < Stage.MAX_STAGE_FAILURES) {
+        assert(scheduler.runningStages.nonEmpty)
+        assert(!ended)
+      } else {
+        // Stage has been aborted and removed from running stages
+        assertDataStructuresEmpty()
+        sc.listenerBus.waitUntilEmpty(1000)
+        assert(ended)
+        assert(jobResult.isInstanceOf[JobFailed])
+      }
+    }
+  }
+
+  /**
+   * In this test we simulate a job failure where there are two failures in two different stages.
+   * Specifically, stage0 fails twice, and then stage1 twice. In total, the job has had four
+   * failures overall but not four failures for a particular stage, and as such should not be
+   * aborted.
+   */
+  test("Failures in different stages should not trigger an overall abort") {
+    val shuffleMapRdd = new MyRDD(sc, 2, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, null)
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep))
+    submit(reduceRdd, Array(0, 1))
+
+    // In the first two iterations, Stage 0 succeeds and stage 1 fails. In the next two iterations,
+    // stage 0 fails.
+    for (attempt <- 1 to Stage.MAX_STAGE_FAILURES) {
+      // Complete all the tasks for the current attempt of stage 0 successfully
+      val stage0Attempt = taskSets.last
+
+      // Confirm  that this is the first attempt for stage 0
+      checkStageId(0, attempt, stage0Attempt)
+
+      if (attempt < Stage.MAX_STAGE_FAILURES/2) {
+        // Make each task in stage 0 success
+        val completions = stage0Attempt.tasks.zipWithIndex.map{ case (task, idx) =>
+          (Success, makeMapStatus("host" + ('A' + idx).toChar, 2))
+        }.toSeq
+
+        // Run stage 0
+        complete(stage0Attempt, completions)
+
+        // Now we should have a new taskSet, for a new attempt of stage 1.
+        // We will have one fetch failure for this task set
+        val stage1Attempt = taskSets.last
+        checkStageId(1, attempt, stage1Attempt)
+
+        val stage1Successes = stage1Attempt.tasks.tail.map { _ => (Success, 42)}
+
+        // Run Stage 1, this time with a task failure
+        complete(stage1Attempt,
+          Seq((FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null))
+            ++ stage1Successes
+        )
+      } else {
+        val stage0Successes = stage0Attempt.tasks.tail.map { _ => (Success, 42)}
+
+        // Run stage 0 and fail
+        complete(stage0Attempt,
+          Seq((FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null))
+            ++ stage0Successes
```

Comment:
I know this really sucks, but I think this test has to be 3 stages as well.  Stage 0 *can't* have a fetch failure, because its not doing any shuffle fetching.

Maybe this kinda works and exercises the right branches, but its hard to guarantee it'll stay that way, so I think we need to create the right stages.